### PR TITLE
recompute priority Api for context

### PR DESCRIPTION
**Pull request by pratikmishra356**
## Problem
After any dimension update, contexts priority needs to be updated

## Solution
Implemented Api inside context handler

## Environment variable changes
NA

## Pre-deployment activity
NA

## Post-deployment activity
NA

## API changes
| Endpoint        | Method           | Request body  | Response Body |
| ------------- |:-------------:| -----:| ----------------:|
| {host}/context/context/priority/recompute    | PUT | NA | { "message": "Contexts priority updated successfully"} |

## Possible Issues in the future
NA

### CODE REVIEW

File: crates/context_aware_config/src/api/context/handlers.rs [hunk 1]
```diff
@@ -526,3 +530,48 @@ async fn bulk_operations(
     })?;
     Ok(Json(response))
 }
+
+#[put("/priority/recompute")]
+async fn priority_recompute(
+    db_conn: DbConnection,
+    _user: User,
```
Comment (Datron) [inline crates/context_aware_config/src/api/context/handlers.rs:537]: Don't take user if you don't want to use it

File: crates/context_aware_config/src/api/context/handlers.rs [hunk 2]
```diff
@@ -526,3 +530,48 @@ async fn bulk_operations(
...
+    let dimension_schema_map = get_all_dimension_schema_map(&mut conn)?;
+    let update_contexts: Vec<Context> = result
+        .iter()
+        .map(|context| Context {
+            priority: calculate_context_priority(
+                "context",
+                &context.value,
+                &dimension_schema_map,
+            )
+            .unwrap(),
```
Comment (Datron) [inline crates/context_aware_config/src/api/context/handlers.rs:556]: No raw unwraps please

File: crates/context_aware_config/src/api/context/handlers.rs [hunk 3]
```diff
@@ -526,3 +530,48 @@ async fn bulk_operations(
...
+    let insert = diesel::insert_into(contexts)
+        .values(&update_contexts)
+        .on_conflict(id)
+        .do_update()
+        .set(priority.eq(excluded(priority)))
+        .execute(&mut conn);
+
+    match insert {
+        Ok(_) => Ok(HttpResponse::Ok()
+            .json(json!({"message": "Contexts priority updated successfully"}))),
```
Comment (Datron) [inline crates/context_aware_config/src/api/context/handlers.rs:569]: Can we send back a response like this:
```
[
    {
        context: String
        old_priority: some value
        new_priority: some value
    },
    {
        next_context
    },
    .........
]
```

**Review by Datron — CHANGES_REQUESTED** [review]

File: crates/context_aware_config/src/api/context/handlers.rs (hunk 3 above)
Comment (pratikmishra356) [reply crates/context_aware_config/src/api/context/handlers.rs:569]: @Datron thinking why this response would be necessary, its a batch operation with many changes , giving an one time response does not make much sense here I think

Comment (Datron) [reply crates/context_aware_config/src/api/context/handlers.rs:571]: @pratikmishra356 Its useful for shell commands and to let the user know what has changed.

Comment (pratikmishra356) [reply crates/context_aware_config/src/api/context/handlers.rs]: done

File: crates/context_aware_config/src/api/context/handlers.rs [hunk 4]
```diff
@@ -526,3 +530,57 @@ async fn bulk_operations(
...
+        log::error!("failed to fetch contexts with error: {}", err);
+        db_error!(err)
+    })?;
+
+    let dimension_schema_map = get_all_dimension_schema_map(&mut conn)?;
+
+    let mut update_contexts: Vec<Context> = vec![];
+    let mut response: Vec<PriorityRecomputeResponse> = vec![];
+
+    for context in result.iter() {
```
Comment (mahatoankitkumar) [inline crates/context_aware_config/src/api/context/handlers.rs:551]: use map to populate `response` and inside that you can push `update_contexts`.

File: crates/context_aware_config/src/helpers.rs [hunk 5]
```diff
@@ -253,6 +253,40 @@ pub fn json_to_sorted_string(v: &Value) -> String {
...
+pub fn calculate_context_priority(
+    object_key: &str,
+    cond: &Value,
+    dimension_schema_map: &HashMap<String, (JSONSchema, i32)>,
+) -> Result<i32, String> {
+    let get_priority = |key: &String, val: &Value| -> Result<i32, String> {
+        if key == "var" {
+            let dimension_name = val
+                .as_str()
+                .ok_or_else(|| "Dimension name should be of String type")?;
```
Comment (mahatoankitkumar) [inline crates/context_aware_config/src/helpers.rs:265]: Better message here.

File: crates/context_aware_config/src/helpers.rs [hunk 6]
```diff
@@ -253,6 +253,40 @@ pub fn json_to_sorted_string(v: &Value) -> String {
     }
 }
 
+pub fn calculate_context_priority(
+    object_key: &str,
+    cond: &Value,
+    dimension_schema_map: &HashMap<String, (JSONSchema, i32)>,
+) -> Result<i32, String> {
+    let get_priority = |key: &String, val: &Value| -> Result<i32, String> {
```
Comment (mahatoankitkumar) [inline crates/context_aware_config/src/helpers.rs:261]:
```suggestion
    let get_priority = |key: &str, val: &Value| -> Result<i32, String> {
```

**Review by mahatoankitkumar — CHANGES_REQUESTED** [review]

File: postman/cac/Context/Recompute Priority Context/response.json [hunk 7]
```diff
@@ -0,0 +1 @@
+[]
```
Comment (Datron) [inline postman/cac/Context/Recompute Priority Context/response.json:1]: delete this file

**Review by Datron — APPROVED** [review]

**Review by mahatoankitkumar — APPROVED** [review]

File: crates/context_aware_config/src/api/context/handlers.rs [hunk 8]
```diff
@@ -526,3 +530,68 @@ async fn bulk_operations(
     })?;
     Ok(Json(response))
 }
+
+#[put("/priority/recompute")]
+async fn priority_recompute(
+    db_conn: DbConnection,
+) -> superposition::Result<HttpResponse> {
```
Comment (ShubhranshuSanjeev) [inline crates/context_aware_config/src/api/context/handlers.rs:537]: The return type could have been `superposition::Result<Vec<PriorityRecomputeResponse>>`

File: crates/context_aware_config/src/api/context/handlers.rs [hunk 9]
```diff
@@ -526,3 +530,68 @@ async fn bulk_operations(
...
+                Err(e) => Err(e),
+            }
+        })
+        .collect::<superposition::Result<Vec<Context>>>()?;
+
+    let insert = diesel::insert_into(contexts)
+        .values(&update_contexts)
+        .on_conflict(id)
+        .do_update()
+        .set(priority.eq(excluded(priority)))
```
Comment (ShubhranshuSanjeev) [inline crates/context_aware_config/src/api/context/handlers.rs:585]: Can you explain little bit, how excluded works?

**Pull request merged**